purchase_order_approval_block/models/purchase_order.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Eficent Business and IT Consulting Services S.L.
+# Copyright 2017 Serpent Consulting Services Pvt. Ltd.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
+
+from odoo import api, fields, models, _
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    approval_block_id = fields.Many2one(
+        comodel_name='purchase.approval.block.reason',
+        string='Approval Block Reason',
+    )
+    approval_blocked = fields.Boolean('Approval Blocked',
+                                      compute='_compute_approval_blocked')
+
+    @api.multi
+    def _compute_approval_blocked(self):
+        for rec in self:
+            if rec.approval_block_id:
+                rec.approval_blocked = True
+
+    @api.model
+    def create(self, vals):
+        po = super(PurchaseOrder, self).create(vals)
+        if 'approval_block_id' in vals and vals['approval_block_id']:
+            po.message_post(body=_('Order \"%s\" blocked with reason'
+                                   ' \"%s\"') % (po.name,
+                                                 po.approval_block_id.name))
+        return po
+
+    @api.multi
+    def write(self, vals):
+        res = super(PurchaseOrder, self).write(vals)
+        for po in self:
+            if 'approval_block_id' in vals and vals['approval_block_id']:
+                po.message_post(
+                    body=_('Order \"%s\" blocked with reason \"%s\"') % (
+                        po.name, po.approval_block_id.name))
+            elif 'approval_block_id' in vals and not vals['approval_block_id']:
+                po.message_post(
+                    body=_('Order \"%s\" approval block released.') % po.name)
+        return res
+
+    @api.multi
+    def button_approve(self, force=False):
+        for rec in self:
+            if rec.approval_block_id:
+                rec.button_release_approval_block()
+        return super(PurchaseOrder, self).button_approve(force=force)
+
+    @api.multi
+    def button_release_approval_block(self):
+        for order in self:
+            order.with_context(
+                force_po_approval_block_release=True).approval_block_id = False
```

Comment:
Why this value in context as it is not used anywhere ?

If you want to add a parameter, I'd prefer having a function `release_approval_block` which do the assignation to False with the parameter force.

Reply:
This is used here https://github.com/OCA/purchase-workflow/pull/446/files#diff-a03a69efc043be23d1167165b82a60ebR25

The only option to avoid this would be to add a boolean field 'Unblock Approval forced' in the PO. When a person adds a block again, this 'Unblock Approval forced' would be reset. Then we would be able to follow your recommendation, but the drawback could be that we store an extra field. On the other hand, perhaps it is a good idea for traceability purposes that you know what Orders were eventually approved, where the block was released under the forced unblock button.

If you want to further clarify the use of the modules we can jump to a Hangouts.

Reply:
I'm ok with this but you have to put with_context on self:

`for order in self.with_context(...):`

My remark was mainly on the new environment creation at each iteration

Reply:
I've applied the changes. However, doing some research about with_context(), I don't see that it is really creating a new environment every time: https://www.odoo.com/documentation/saas-13/reference/orm.html

I'm happy with the code now, could you approve the changes?

Reply:
Yes I'll do it

That's where Odoo is tricky with its documentation!

https://github.com/odoo/odoo/blob/8165018058f59eedecc2fa0b9bde93e505d56c53/odoo/models.py#L4890

Reply:
Okay, I see now. Thank you for your feedback. I'll keep this in mind! :+1: